### PR TITLE
Color Block Support: Switch to ToolsPanel for displaying UI

### PR DESCRIPTION
**Pull request by aaronrobertshaw**
Depends on:
- https://github.com/WordPress/gutenberg/pull/37546

Related: 
- https://github.com/WordPress/gutenberg/issues/27331
- https://github.com/WordPress/gutenberg/pull/33744
- https://github.com/WordPress/gutenberg/pull/33743

## Description

This updates the UI for controlling colors provided by block supports to use the `ToolsPanel` component.

- With the adoption of `ToolsPanel` for other block supports, this aims to improve consistency between the block support panels.
- Introduces a new `ToolsPanelColorDropdown` component. 
    - Prior to this switch of panel, color controls are rendered within an `ItemGroup`. 
    - For us to allow ad hoc color controls to be injected into this panel by blocks, we cannot use the `ItemGroup`. 
    - Controls are injected via slot fills and to maintain their visual order when not toggled on for display, hidden placeholder elements must be rendered which prevents appropriate `ItemGroup` styling.
    - In addition, if the panel itself were made an `ItemGroup` we could end up with non-color controls as children of the `ItemGroup` breaking its semantics (e.g. Contrast Checkers, help text, or anything else a 3rd party dev wishes to render into the slot)
    - Instead, this PR simply replicates the styling of the `ItemGroup` and leverages CSS ordering to force contrast checkers to the bottom of the panel.
- The Global Styles sidebar uses different navigation and display of color options so this PR only updates the panel used in the inspector controls sidebar of the block editor.

_Note: Which color controls display by default (even without any value) for any given block will be assessed and updated via a separate PR. This only focuses on switching the panel component the controls are within._

## How has this been tested?
Manually.

#### Test Instructions
1. Create a new post, add a group block, and select it
2. In the inspector controls sidebar you should see the colors block support panel using the `ToolsPanel`. It will have a `+` icon for the "more" menu
3. Explore toggling color controls and updating their values, making sure they still function appropriately
4. Try saving color values, reloading the editor, and ensuring that selected colors expose those controls by default

## Screenshots 

https://user-images.githubusercontent.com/60436221/146899790-8d6a8d0d-fcb5-49a4-83d8-614af0955d15.mp4

## Types of changes
Enhancement

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

**Comment by jasmussen** [comment]
Thanks for all the tools menu efforts 👌 

This is what I see:

color

So, ultimately I do think we need to make the color panel into a ToolsPanel, notably when we get a slew of color values such as link, link hover, link focus, and others which are important for global styles but should very rarely be customized on, for example, a single Paragraph block.

However specifically for the Paragraph block (and every other block with a color panel I can think of, including Button and Group), I would expect at least text and background color options to be shown by default. 

The problem with color swatches being duplicated and taking up a lot of space still needs a solution such as this one from #27331: 



— that, but updated to the design in https://github.com/WordPress/gutenberg/issues/27473#issuecomment-796773866.

It seems okay in the mean time to convert to the ToolsPanel, but still showing at least text and background by default. Happy to note that all supported colors are already shown in Global Styles, as they should be 👌 

Question: do you have a list of blocks that are affected by this PR? And can we customize the defaults per-block type yet?

**Comment by apeatling** [comment]
We are holding off on this conversion until we have better color controls in place as talked about in https://github.com/WordPress/gutenberg/issues/27331

129268741-bfa7b547-223e-49a7-9388-f4f35bafa2ce

**Comment by aaronrobertshaw** [comment]
This PR has been rebased and updated with respect to the latest changes to the `ToolsPanel`. It also includes a new `color` group for the InspectorControls slots to inject controls into this panel.

Hopefully, when the new color controls are ready there is minimal extra work required to finish this switch off.

File: packages/block-editor/src/hooks/color.scss
Comment (ciampo) [inline packages/block-editor/src/hooks/color.scss:9]: Is this change necessary after the latest changes to `ToolsPanel`'s layout?

**Review by ciampo — COMMENTED** [review]
Hey @aaronrobertshaw , is this PR in need of a review? I can give it a deeper look after the required rebase, in case you need !

File: packages/block-editor/src/hooks/color.scss
Comment (aaronrobertshaw) [reply packages/block-editor/src/hooks/color.scss:9]: The contrast checker [isn't added as a `ToolsPanelItem`](https://github.com/WordPress/gutenberg/blob/update/color-support-to-use-tools-panel/packages/block-editor/src/hooks/color-panel.js#L89-L94) and as such doesn't get the [full-width style](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/tools-panel/styles.ts#L22-L24) by default.  This could possibly be removed when we settle on what `Grid` or `GridItem` props to expose for the `ToolsPanel`.

**Comment by aaronrobertshaw** [comment]
> Is this PR in need of a review? I can give it a deeper look after the required rebase, in case you need !

Thanks @ciampo. I've rebased the PR and fixed some issues after recent changes to the color control displaying multiple palettes from different origins. 

I ran short on time today so I haven't yet had the chance to test this fully or polish the code changes. I don't think there is a great hurry to merge so you're welcome to hold off on reviewing until later in the week if that suits better.

File: packages/block-editor/src/hooks/color.scss
Comment (ciampo) [reply packages/block-editor/src/hooks/color.scss:9]: > The contrast checker isn't added as a ToolsPanelItem

I'm probably missing something, but why can't it be added as a `ToolsPanelItem`?

**Review by ciampo — COMMENTED** [review]
I followed the testing instructions and found a few differences:

> In the inspector controls sidebar you should see the colors block support panel using the ToolsPanel. It will have a ... more menu

In my tests, the `Color` panel has a `+` icon for the more menu, which changes into a `⋮` icon when at least one `ToolsPanelItem` is visible (_I believe this is working as expected and that the testing instructions may be out of date, but I thought I'd flag it anyway_)

> Explore toggling controls and updating color values, making sure they still function appropriately

In my tests, the `Link Color` doesn't seem to make a difference: in the example screenshot, I set the link color to be taupe, but the link appears as black.

Screenshot 2021-11-11 at 19 34 25

> Switch to the site editor and confirm the display and function of the color panel under general global styles and the By block type tab

Not sure about what this means exactly — could you be more specific?

**Comment by aaronrobertshaw** [comment]
> In my tests, the Color panel has a + icon for the more menu, which changes into a ⋮ icon when at least one ToolsPanelItem is visible (I believe this is working as expected and that the testing instructions may be out of date, but I thought I'd flag it anyway)

You're correct, the instructions pre-dated the change in icon, which landed in https://github.com/WordPress/gutenberg/pull/34107. 

> > Switch to the site editor and confirm the display and function of the color panel under general global styles and the By block type tab
> 
> Not sure about what this means exactly — could you be more specific?

You can ignore that previous testing instruction. The global styles sidebar navigation and the color controls within it have changed recently and so the old color panel used there doesn't exist anymore.

I've updated the PR description and testing instructions so hopefully they are clearer now. Sorry about that.

> In my tests, the Link Color doesn't seem to make a difference: in the example screenshot, I set the link color to be taupe, but the link appears as black.

I also re-recorded the demo video to test the link color issue. I couldn't replicate the problem you described. The link color still works for me in multiple themes. Can you share any other details that might help track that down?

| 2022 | 2021 |
|---|---|
| Screen Shot 2021-11-12 at 1 18 11 pm | Screen Shot 2021-11-12 at 1 49 42 pm |

**Comment by ciampo** [comment]
> Can you share any other details that might help track that down?

To set up:

1. Checkout branch locally
2. Run `npm run distclean && npm ci && npm run wp-env start && npm run dev`
3. Once all processes have finished running, visit `localhost:8888/wp-admin` and login to the dashboard


Then I took a video:

https://user-images.githubusercontent.com/1083581/141461482-c3f3290e-cdfb-4037-a42f-d6539b1610a5.mp4


Notice how the link color changes unexpectedly when changing the background color, and how it doesn't follow the "link color" set via the Tools Panel.

I then switched theme to TT1-block (from Twenty Twenty One), and the behaviour seems more coherent:

https://user-images.githubusercontent.com/1083581/141462365-71fb8723-ddea-43bb-8a13-a9a9d4fb2fc5.mp4


Which makes me think that this bug may not be related to this PR necessarily — but definitely something we need to look at.

**Comment by aaronrobertshaw** [comment]
Thanks for the extra info and for taking the time to record the videos @ciampo. 🙇 

This does appear to be a theme-specific issue. I can replicate this behaviour on `trunk` with the TwentyTwentyOne classic theme. It [has styles](https://core.trac.wordpress.org/browser/trunk/src/wp-content/themes/twentytwentyone/style.css#L1716) overriding link color in a number of places. 

I have a recollection this was a known issue but I couldn't find a relevant ticket after a quick search. I believe the block support or theme.json that generates these color styles can't increase their specificity since that would negatively impact every other theme. It will need to be addressed within TwentyTwentyOne classic theme itself.

All that said, I don't think that specific issue should block this PR.

File: packages/block-editor/src/hooks/color.scss
Comment (aaronrobertshaw) [reply packages/block-editor/src/hooks/color.scss:9]: > I'm probably missing something, but why can't it be added as a ToolsPanelItem?

Adding it as a `ToolsPanelItem` means it would be displayed within the `ToolsPanel` menu. It isn't something that the user should be toggling on or off. 

If the contrast checker is wrapped up inside a `ToolsPanelItem` with another control say background color, it wouldn't appear in the menu but it wouldn't show when it should. Take the scenario where the user only toggles on the font color control and not the background color. They change that font color to something that should trigger the warning from the contrast checker however because the contrast checker is wrapped with the background color and that isn't displayed, the contrast checker remains hidden.

Duplicating the contrast checker so one each is wrapped in the color controls would mean when both font and background color are toggled on for display the user would get duplicate warnings. 

I believe this simple CSS adjustment for something that is related specifically to this panel should be ok.

Comment (ciampo) [reply packages/block-editor/src/hooks/color.scss:9]: Thank you for the explanation, now I understand the situation much better.

> I believe this simple CSS adjustment for something that is related specifically to this panel should be ok.

That definitely sounds sensible for this PR!

I just didn't want to miss an interesting opportunity to test potential limitations of the current `ToolsPanel` component — do you think we could improve `ToolsPanel` to handle this situation better? (again, to be tackled in a separate PR)

**Review by ciampo — APPROVED** [review]
> All that said, I don't think that specific issue should block this PR

Agreed — thank you for the detailed explanation confirming that what I reported is not a bug related to this PR.

I believe we can ho ahead and merge this PR — the ongoing conversation regarding the contrast checker not being wrapped in a `ToolsPanelItem` should not be considered blocking either, and any outcome may be tackled separately.

File: packages/block-editor/src/hooks/color.scss
Comment (aaronrobertshaw) [reply packages/block-editor/src/hooks/color.scss:9]: > do you think we could improve ToolsPanel to handle this situation better?

No great ideas come to mind. 

In terms of styling, we either need to:
- add some generic default that might not apply in all cases to the `ToolsPanel`, 
- stuff the `ToolsPanel` full of styles for random components,
- add a style for ad hoc components ( as done here ), 
- or, wrap it in a component that provides that style or equivalent props e.g. `grid-column`.

When switching the `ToolsPanel` to use the `Grid` component to handle its layout, it was decided we shouldn't expose extra props or add a `GridItem` like component. I would expect a case like this to be covered when that decision gets revisited.

Comment (ciampo) [reply packages/block-editor/src/hooks/color.scss:9]: > When switching the ToolsPanel to use the Grid component to handle its layout, it was decided we shouldn't expose extra props or add a GridItem like component. I would expect a case like this to be covered when that decision gets revisited.

Yeah, that is definitely clear. I think that adding a style for ad hoc components ( as done here ) is the best choice at the moment.

---

The aspect that I was highlighting here is that we are displaying a component in the `ToolsPanel`  (the contrast checker) that can't be wrapped in a `ToolsPanelItem`, since it needs to be always visible (and should not appear in the `ToolsPanel` menu) — I was wondering if this feels hacky, but after reviewing how `ToolsPanel` is supposed to work, it looks like a perfectly fine pattern. At most, we could add a similar case scenario in the `ToolsPanel` readme to explain that items that are not wrapped in `ToolsPanelItem` are still going to be rendered as expected.

Definitely not a blocker for this PR in any case!

Comment (aaronrobertshaw) [reply packages/block-editor/src/hooks/color.scss:9]: >  I was wondering if this feels hacky, but after reviewing how ToolsPanel is supposed to work, it looks like a perfectly fine pattern

This aspect of the `ToolsPanel` design was intended and was suggested as feedback in some of the earliest iterations of the component. 

From memory, the earliest incarnation converted all immediate panel children to items in the menu. The shift to allowing panel items to register themselves opened up the possibility that not everything had to be included in the menu. 

My thinking at the time, was that we might have intro or summary text in a panel. Something similar to some of the Global Styles panels. I wasn't anticipating the display of those elements in the panel to be related to the items that were controlled via the menu but it's not a big leap.

Screen Shot 2021-11-17 at 4 56 30 pm

> At most, we could add a similar case scenario in the ToolsPanel readme to explain that items that are not wrapped in ToolsPanelItem are still going to be rendered as expected.

I'm happy to update the `ToolsPanel` readme and stories to include something that demonstrates that items not wrapped in a `ToolsPanelItem` are going to be rendered and is expected.

Comment (aaronrobertshaw) [reply packages/block-editor/src/hooks/color.scss:9]: PR improving the `ToolsPanel` readme and stories is available in: https://github.com/WordPress/gutenberg/pull/36557

Comment (ciampo) [reply packages/block-editor/src/hooks/color.scss:9]: Thank you for the explanation and additional context, it sounds good to me — we can keep things as they currently are. I just wanted to make sure that the way this worked was "expected" and I lacked the context that you shared!

We can mark this conversation as solved :)

**Comment by aaronrobertshaw** [comment]
This PR has been rebased and updated to pick up and use the latest color controls (i.e. the color dropdowns). For this to work, it required some tweaks to the `ToolsPanel` which have been made in a separate PR (https://github.com/WordPress/gutenberg/pull/37546).

The color block support inspector controls slot was also tweaked so that it will render the slot container as an `ItemGroup`. The individual `ToolsPanelItem` components for the color support now render as `Item` components as well to maintain the approach taken in the current color controls on `trunk` as best possible.

To assist future injection of ad hoc color controls for individual blocks into the colors panel, I've extracted the `ToolsPanelItem` generation into its own component so that consumers will only need to pass it settings similar to other color components, just with the addition of those props required by the `ToolsPanelItem`.

The PR description has been updated to include a new demo video.

File: packages/block-editor/src/hooks/color.scss
Comment (ramonjd) [inline packages/block-editor/src/hooks/color.scss:45]: Given the depth, it's probably rare that these styles will be overwritten (in the absence of !important) but do you think it's worth making the `first` and `last` classnames more specific, e.g., `components-tools-panel-item__first`?

Comment (ciampo) [reply packages/block-editor/src/hooks/color.scss:45]: See related conversation https://github.com/WordPress/gutenberg/pull/37546#discussion_r779740754

**Comment by aaronrobertshaw** [comment]
After [discussions](https://github.com/WordPress/gutenberg/pull/37546#discussion_r784979905) on the [PR adding first and last visible item classes](https://github.com/WordPress/gutenberg/pull/37546) within a ToolsPanel, it was raised that with the current method of adding color controls to the grouped inspector controls panel, we could have non-`Item` elements added and the panel's slot wrapper can't then be an `ItemGroup`.

The simplest option was to remove the use of `ItemGroup` for the color panel but still style the color options in the same manner.

If it is still desired that all color options appear within an `ItemGroup` with the corresponding `role` attributes, we might need to look at creating additional nested slots. Those are in turn rendered into the current color panel slot. This would make the color panel different from all the other grouped inspector control slots and we'd probably need to enforce only the new slots were rendered within the current one. I think this could be done as a separate follow up if desired.

**Comment by ciampo** [comment]
> The simplest option was to remove the use of `ItemGroup` for the color panel but still style the color options in the same manner.

Yep, this is the direction to follow (at least for now). We may just want to have @jasmussen to check the border styles (e.g. border color, radius..) to make sure it fits our design guidelines.

> we might need to look at creating additional nested slots. Those are in turn rendered into the current color panel slot. This would make the color panel different from all the other grouped inspector control slots and we'd probably need to enforce only the new slots were rendered within the current one. I think this could be done as a separate follow up if desired.

In case that was related to using `ItemGroup`, it can be dropped for now (as we are not going to use `ItemGroup`, at least in the short term)

**Comment by aaronrobertshaw** [comment]
> In case that was related to using ItemGroup, it can be dropped for now (as we are not going to use ItemGroup, at least in the short term)

Yes, I meant this possible course of action as an option only if we decided to enforce use of `ItemGroup` around just the color options. I think if that is decided it makes sense as a separate follow up.

I'll aim to merge the PR adding the necessary first/last visible item classes to the `ToolsPanel` today and rebase this to absorb those changes. At that point, I think we'll be ready to pull the trigger on landing this pending final approval.

**Comment by ramonjd** [comment]
This is testing well for me

✅ In the inspector controls sidebar you should see the colors block support panel using the ToolsPanel. It will have a + icon for the "more" menu
✅ Explore toggling color controls and updating their values, making sure they still function appropriately
✅ Try saving color values, reloading the editor, and ensuring that selected colors expose those controls by default


I did notice that, on the frontend, a container's link color overrides the link color of its inner blocks, but it's happening on trunk as well. Just noting here. I'll create an issue later if none exists.

Screen Shot 2022-02-08 at 1 48 13 pm


Screen Shot 2022-02-08 at 1 49 04 pm

**Review by ramonjd — APPROVED** [review]
Nice work! Works as advertised for me. 🙇

**Comment by aaronrobertshaw** [comment]
> I did notice that, on the frontend, a container's link color overrides the link color of its inner blocks, but it's happening on trunk as well. Just noting here. I'll create an issue later if none exists.

Nice catch. I agree we should tackle that one separately as I don't believe it is directly related to this PR as you noted.

**Comment by Mamaduka** [comment]
Great work! 🦸

File: packages/block-editor/src/components/colors-gradients/tools-panel-color-dropdown.js
Comment (youknowriad) [inline packages/block-editor/src/components/colors-gradients/tools-panel-color-dropdown.js:24]: Why did we introduce a new component instead of updating the existing `ColorGradientSettingsDropdown`? Now basically we have two concurrent components for developers to use and they have duplicated styles and very similar behavior.

cc @jorgefilipecosta as we discussed this a lot when we first updated `ColorGradientSettingsDropdown` to use the global styles like UI

File: packages/block-editor/src/components/colors-gradients/tools-panel-color-dropdown.js
Comment (youknowriad) [inline packages/block-editor/src/components/colors-gradients/tools-panel-color-dropdown.js:54]: This is reusing a classname from another component potentially creating style conflicts.

File: packages/block-editor/src/components/colors-gradients/tools-panel-color-dropdown.js
Comment (youknowriad) [inline packages/block-editor/src/components/colors-gradients/tools-panel-color-dropdown.js:60]: This is reusing a classname from another component potentially creating style conflicts.

File: packages/block-editor/src/components/block-inspector/index.js
Comment (youknowriad) [inline packages/block-editor/src/components/block-inspector/index.js:149]: This class doesn't seem to follow our naming guidelines. We should start with the package name followed by the component name and a modified, so something like `block-editor-block-inspector__color-slot`

File: packages/block-editor/src/hooks/color.scss
Comment (youknowriad) [inline packages/block-editor/src/hooks/color.scss:31]: Should we move these styles to the component style itself (style.scss of the same folder of the component). Or are these styles specific to this particular usage of this component. If I'm not wrong this component is specific to hooks anyway.

Comment (youknowriad) [reply packages/block-editor/src/hooks/color.scss:31]: (Similar question for the remaining styles here)

File: packages/block-editor/src/hooks/color.scss
Comment (youknowriad) [inline packages/block-editor/src/hooks/color.scss:1]: Where does this style come from, it doesn't seem to follow our style guidelines.

File: packages/block-editor/src/components/colors-gradients/tools-panel-color-dropdown.js
Comment (aaronrobertshaw) [reply packages/block-editor/src/components/colors-gradients/tools-panel-color-dropdown.js:24]: This PR was already open and underway before the color controls got updated. It turned out they were updated without taking into account the future use case where they would be within the `ToolsPanel`. At this point we have two different needs for a color control. One that can be used directly in a limited `ItemGroup` and another that can exist within a `ToolsPanel`.

The `ColorGradientSettingsDropdown` renders a group of color controls within an `ItemGroup`. This can't really be done when the possible color controls are being rendered via the block support SlotFills and to maintain ordering of `ToolsPanel` controls placeholders must be rendered which interferes with `ItemGroup` styling and layout. There are also no restrictions on what can be rendered via those slots meaning it could contain additional help text, contrast checkers and so on.

The earlier comments and conversation on this PR also provide some background as to the issues faced as well as alternative approaches explored such as attempting to make the color block support slot render as an `ItemGroup` etc.

File: packages/block-editor/src/components/colors-gradients/tools-panel-color-dropdown.js
Comment (aaronrobertshaw) [reply packages/block-editor/src/components/colors-gradients/tools-panel-color-dropdown.js:54]: Agreed. To maintain the visual consistency, are you okay with us duplicating all these styles?

File: packages/block-editor/src/components/block-inspector/index.js
Comment (aaronrobertshaw) [reply packages/block-editor/src/components/block-inspector/index.js:149]: I can create a PR to tidy this up.

File: packages/block-editor/src/hooks/color.scss
Comment (aaronrobertshaw) [reply packages/block-editor/src/hooks/color.scss:31]: Currently, the component is only used in the color block supports hook but could be used to add a color control to any `ToolsPanel`. I can move these styles to live in the `color-gradients/style.scss` if that is better.

File: packages/block-editor/src/hooks/color.scss
Comment (aaronrobertshaw) [reply packages/block-editor/src/hooks/color.scss:1]: The [classname comes from the `BlockSupportsToolsPanel`](https://github.com/WordPress/gutenberg/blob/8675f86f885ca389f3bca39971b22cb897bab61b/packages/block-editor/src/components/inspector-controls/block-support-tools-panel.js#L66) which wraps the grouped inspector controls slots for block supports in a `ToolsPanel`.

I can update the format of this class and the corresponding styles for the different block supports dimensions, border, typography etc.

**Comment by aaronrobertshaw** [comment]
Thanks for raising these issues and concerns @youknowriad.

Are you ok with `ToolsPanelColorDropdown` and `ColorGradientSettingsDropdown` remaining separate components given their different usage and requirements? If so, I'll create a PR avoiding re-using the same classes and styles but keeping them visually the same.

I'll also create another PR updating the class name for the block support panels, the corresponding styles, and e2e tests. 

Is there anything else you would like to see fixed up in this area?

File: packages/block-editor/src/components/colors-gradients/tools-panel-color-dropdown.js
Comment (youknowriad) [reply packages/block-editor/src/components/colors-gradients/tools-panel-color-dropdown.js:24]: Can't we just replace the ItemGroup version with this one which if I'm understanding properly is the same visually, but a bit different in terms of component tree in order to allow extensibility?

Comment (aaronrobertshaw) [reply packages/block-editor/src/components/colors-gradients/tools-panel-color-dropdown.js:24]: I'm not sure we can, or at least, not directly.

The `ItemGroup` based version allows for a full set of color controls to be grouped semantically. The separate version added here was for when that wasn't possible i.e. within the context of a `ToolsPanel`. It also allowed for the controls to be rendered individually rather than as part of a group e.g. injecting an individual color control into the color block supports panel.

The original `ItemGroup` version is still in use in the Cover block, for its opacity controls via `PanelColorGradientSettings`, as well as possibly by third party plugins etc. Also, `PanelColorGradientSettings` has a native implementation that I don't think makes sense for the dropdown component introduced in this PR given the `ToolsPanel` is only used on the web.

Apologies if I'm misunderstanding the question.

Comment (youknowriad) [reply packages/block-editor/src/components/colors-gradients/tools-panel-color-dropdown.js:24]: From a user's perspective, both seem to have the same output. Semantics seem internal implementation details. Let me ask the question differently.

From a user perspective, it seems to me like `PanelColorGradientSettings` is a special case (no extensibility version) of  `ToolsPanelColorDropdown`. Can we make the `ToolsPanelColorDropdown` API (props) to match `PanelColorGradientSettings` in which case, there's no impact on third-party or core usage (both functionality wise since it's just an addition of extensibility, and in terms of code since the APIs match).

Maybe I'm missing something here as well :)

Comment (aaronrobertshaw) [reply packages/block-editor/src/components/colors-gradients/tools-panel-color-dropdown.js:24]: > Maybe I'm missing something here as well :)

I think it's more likely confusion on my part.  Thanks for your patience. 🙏 

> From a user's perspective, both seem to have the same output. Semantics seem internal implementation details.

I can agree that visually, the output from a single `ColorGradientSettingsDropdown` passed an array of settings objects will look much the same as a series of individual `ToolsPanelColorDropdown` components rendered in the Color block support `ToolsPanel`. 

That was intended so the color controls within the block support `ToolsPanel` still matched the design / current appearance.

Another point of confusion could be the name of `ColorGradientSettingsDropdown` is misleading as well given it actually renders more than a single dropdown. A better name might be `ColorGradientSettingsItemGroup`.

> From a user perspective, it seems to me like PanelColorGradientSettings is a special case (no extensibility version) of ToolsPanelColorDropdown.

Unfortunately, I'm not following this comment. 

To me they do different things. The `PanelColorGradientSettings` wraps a `ColorGradientSettingsDropdown` in a `PanelBody` so we end up with a fixed standalone panel like:

```
<PanelBody>
  <ItemGroup>
    <Item>
      <Dropdown>
    </Item>
    <Item>
      <Dropdown>
    </Item>
    ...
  </ItemGroup>
</PanelBody>
```

Whereas the `ToolsPanelColorDropdown` only gives us a `ToolsPanelItem` that can be rendered as a fill where desired:

```
<ToolsPanelItem>
    <Dropdown>
</ToolsPanelItem>
```

This allows that to be injected into our color block support panel resulting in:

```
<ToolsPanel>
  <ToolsPanelItem>
      <Dropdown>
  </ToolsPanelItem>
  ...
</ToolsPanel>
```

> Can we make the ToolsPanelColorDropdown API (props) to match PanelColorGradientSettings in which case, there's no impact on third-party or core usage (both functionality wise since it's just an addition of extensibility, and in terms of code since the APIs match).

I could see more of a possibility of aligning the `ColorGradientSettingsDropdown` and the `ToolsPanelColorDropdown`.  This was what I thought you were originally suggesting.

Merging these two components would be achievable but in my view, it overcomplicates the resulting component and its use by consumers. I appreciate that once introduced we need to support this new component so I'm open to refactoring this.

All that said, I acknowledge there is still the problem of the shared styles.

Comment (youknowriad) [reply packages/block-editor/src/components/colors-gradients/tools-panel-color-dropdown.js:24]: Thanks for clarifying more it makes a lot of sense.

> I could see more of a possibility of aligning the ColorGradientSettingsDropdown and the ToolsPanelColorDropdown. This was what I thought you were originally suggesting.

Yeah for me, I was discussing the "whole thing" as well, not a single item so I probably got confused in the naming as well. 

--
Let me try to clarify how I end up here, and I think it will help you understand my concerns and allow us (probably you since you know more than me about these components) find a path forward.

I wanted to ask a button (reset icon) on the right of each item in the color panel. The button would serve as a way to "reset/clear the value of the given item without the need to reach out to the three dots menu. While doing so, I noticed that these color buttons are duplicated in two different components and wasn't sure which one I should be using/updating (probably both), so I thought that there might not be a reason for these "color panel items" to be duplicated.

Comment (aaronrobertshaw) [reply packages/block-editor/src/components/colors-gradients/tools-panel-color-dropdown.js:24]: > I wanted to ask a button (reset icon) on the right of each item in the color panel. The button would serve as a way to "reset/clear the value of the given item without the need to reach out to the three dots menu. 

As you note, the Color panel in the inspector controls sidebar is a `ToolsPanel` with an ellipsis menu. If it is only in this panel, you wish to add the reset button, it could be added to the `ToolsPanelColorDropdown`. I can also see why we'd want to be consistent and add it to the controls when they aren't within a `ToolsPanel` as well.

One thing I'd note here is that a goal for the `ToolsPanel` was to declutter the sidebar removing the repeated reset/clear buttons everywhere. @jasmussen might be better positioned to expand on this. My knowledge on current design goals there could also be out of date.

> so I thought that there might not be a reason for these "color panel items" to be duplicated.

Understandable. Hopefully, the extra information in my earlier comments illustrates that while there's overlap, they serve different needs. 

I can draft a PR that explores merging these components.

Comment (jasmussen) [reply packages/block-editor/src/components/colors-gradients/tools-panel-color-dropdown.js:24]: Thank you for the ping. Yes indeed, that's still a task that needs careful thought. One first step I've explored is a reset button on the itemgroups, like Riad suggests:

Clearing

This would share some DNA with ideas outlined in #39427.

A quicker/simpler fix might be to re-introduce the reset button inside the color popover. @aaronrobertshaw I recall you had one in the border popover, right? As we explore options to clarify the whole flow, a simple clear button like that inside all color popovers might be fine.

I'd make it a tertiary not-small style button though. What do you think?

Comment (aaronrobertshaw) [reply packages/block-editor/src/components/colors-gradients/tools-panel-color-dropdown.js:24]: > I'd make it a tertiary not-small style button though. What do you think?

Thanks for the designs @jasmussen, that makes sense now. I'm not opposed to the unobtrusive tertiary button described. Anything that helps sort out the UX around resetting values is fine by me 👍 

> A quicker/simpler fix might be to re-introduce the reset button inside the color popover. @aaronrobertshaw I recall you had one in the border popover, right?

There is a "Reset to default" button in the new `BorderControl`'s color and style picker that will clear the border color and style values.

Comment (aaronrobertshaw) [reply packages/block-editor/src/components/colors-gradients/tools-panel-color-dropdown.js:24]: I've created a PR (https://github.com/WordPress/gutenberg/pull/40084) to merge the two color dropdown related components. I'll resolve this discussion so it can be continued there with tangible code changes to review.

Comment (jasmussen) [reply packages/block-editor/src/components/colors-gradients/tools-panel-color-dropdown.js:24]: Thanks Aaron! 

And just to be clearery about the clear button, I was thinking that instead of this:
Screenshot 2022-04-06 at 09 58 00

we do this:

Screenshot 2022-04-06 at 09 57 43

**Pull request merged**
